review-problems-sol.ml

```diff
@@ -212,6 +219,41 @@ let poly_gen' cs =
   in
   go (fun _ -> 1) cs
 
+(* Binary search tree problems *)
+(* Definitions *)
+type 'a tree = 
+  |Empty 
+  |Node of ('a tree * 'a * 'a tree)
+ 
+type 'a bst = (int * 'a) tree (* Key-Value pair*)
+
+
+(* insert : 'a bst -> int -> 'a -> 'a bst *)
+(* Not tail-recursive *)
+let rec insert (tree:'a bst) (key: int) (value: 'a) :'a bst = 
+  match tree with
+  |Empty -> Node (Empty, (key, value), Empty)
+  |Node (l,(k,v), r) -> 
+      if k = key then (Node (l, (k, value),r)) else
+      if k > key then (Node (insert l key value, (k,v), r)) else 
+        (Node (l,(k,v),insert r key value)) 
+
+let rec lookup tree key = 
+  match tree with
+  |Empty -> None
+  |Node (l,(k,v), r) -> 
+      if k = key then Some v else
+      if k > key then lookup l key else
+        lookup r key
+
+let rec insert_gen tree key = 
+  match tree with
+  |Empty -> (fun x -> Node (Empty, (key, x), Empty))
+  |Node (l,(k,v), r) -> 
+      if k = key then (fun x -> (Node (l, (k, x),r))) else
+      if k > key then (fun x -> (Node ((insert_gen l key) x, (k,v), r))) else 
+        (fun x -> (Node (l,(k,v),(insert_gen r key) x)))
```

Comment:
This solution is incorrect.
It's tricky to get these generator problems right.
In short, the functions that get returned must not contain any recursive calls in them. All the recursion has to happen up front. To solve this, hoist the recursive calls out of the returned functions and place them into `let-in`s. 
For example,
```ocaml
if k > key then
  let f = insert_gen l key in
  fun x -> Node (f x, (k, v), r)
else
  ...
```
This way, all the recursion happens right away, and the function that's returned ends up taking the form of a linked list (notice how each returned function refers to the result of the recursive call `f`). When the ultimately returned function is called, it places the element `x` with two empty subtrees beside it, and then rebuilds the whole tree starting from there.

Reply:
Hmmm, this one I have to think for a bit...

Reply:
Sorry for the wait! My brain was on low-battery for the past few days, but I corrected the `insert_gen` function. 

I think I sort of get it now, the correct version generates the tree from "bottom-up" (It computes the bottom-most node when given an input and pass that node all the way to the top until the tree is built) whereas my original one is "top-down". (it searches down for the right spot only when given an input) 

I think the correct one is actually a bit more efficient simply because the traversal is only done once, whereas the original one has to traverse the tree every time it is given an input. (I think that's the case?)

I can certainly see value in writing it this way then, if that's the case; especially if the given tree is some slow to traverse n-th ary tree with a tons of node. (Yes, I just used my prof as a rubber duck to undertand the code :p)

Reply:
The recursion is the same in both ways of writing it. The difference is in _when_ the recursion happens. There are two distinct phases in this recursive algorithm:
1. Going down the tree to find the spot to insert the new key-value pair.
2. Going back up and reconstructing the tree.
The correct implementation of `insert_gen` splits up these two phases. `insert_gen` performs _only_ the first phase (figuring out where to put the key-value pair). The function that `insert_gen` returns performs _only_ the second phase (reconstructing the tree).

To see these two phases, consider this excerpt from the original (non-generator) algorithm:
```ocaml
  | Node (l, (k, v), r) -> (* ... *) Node (insert (k, v) l, x, r)
```

The order of operations is:
1. Make a recursive call to `insert` on the left subtree
2. Reconstruct the tree with the result of the recursive call on the left.

In the recursive call however, the same thing will happen again until we reach an `Empty` tree, so a stack of pending "reconstruct the tree with the result of the recursive call" operations ends up being built up.
In the generator code, this stack of pending operations is explicitly formed as the chain of functions that gets returned by the generator.

The benefit of this split of the phases of the recursion is that we can potentially call the returned function multiple times. In this case, it would allow us to construct many slightly different trees, where we vary the element that gets inserted, but it gets inserted into the same place.